{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4ec1c110-558f-4f76-b385-30525c7774df",
   "metadata": {},
   "source": [
    "### Get 16S sequence embedded into a model,mayneed to not use just a general language model honestly tho...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9255ae5e-d982-4e59-9cfb-281e6f0d9a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sentence_transformers import SentenceTransformer, util\n",
    "import pandas as pd\n",
    "from transformers import AutoTokenizer, AutoModel\n",
    "import numpy as np\n",
    "import torch\n",
    "import random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e82ef3e5-ea91-408a-830f-6da835006280",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Single list of sentences - Possible tens of thousands of sentences\n",
    "sentences = [\n",
    "    \"TGCTTCTTGGGTGTTGAGCGTGACC\",\n",
    "    \"GTGATCCCTTCGTAGCACCTGGGAA\",\n",
    "    \"GAGCTCGAATAGGTTATCAGAGCAG\",\n",
    "    \"CGACCGATGTCAAGAGCGCGCGCGT\",\n",
    "    \"TCTTACCTACTTGACTCTAGTGCGG\",\n",
    "    \"CCGACTGTTAGTCTCAGTTGTACTA\",\n",
    "    \"TGCGAACGTGGACCTGGTGCGAAGC\",\n",
    "    \"AGCTTGCGTGGATTCGTAGGGCTGG\",\n",
    "    \"CGTCCCGTTGCATATCACGAGAGCG\",\n",
    "    \"GGTTACACGTCTACTACTACATCCG\"]\n",
    "\n",
    "\n",
    "paraphrases = util.paraphrase_mining(model, sentences)\n",
    "\n",
    "for paraphrase in paraphrases[0:10]:\n",
    "    score, i, j = paraphrase\n",
    "    print(\"{} \\t\\t {} \\t\\t Score: {:.4f}\".format(sentences[i], sentences[j], score))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d170a6c2-cd6c-4d9f-ab8c-ec92bb06202a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                             sequence\n",
      "id                                                                   \n",
      "GB_GCA_009371985.2  GAAGAGTTTGATCCTGGCTCAGGATGAACGCTAGCGGCAGGCCTAA...\n",
      "RS_GCF_902705855.1  GGAGGTGATCCAGCCCCAGGTTCCCCTAGGGCTACCTTGTTACGAC...\n",
      "RS_GCF_000426325.1  AGGAGGTGATCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...\n",
      "RS_GCF_004343615.1  TGAGAGTTTGATCCTGGCTCAGAGCGAACGCTGGCGGCATGCTTAA...\n",
      "RS_GCF_010727475.1  AGGAGGTGATCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...\n",
      "RS_GCF_006540305.1  GGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAA...\n",
      "RS_GCF_002022265.1  GCAACCCTCTGTACCGACCATTGTAGCACGTGTGTAGCCCAGGCCG...\n",
      "GB_GCA_016192685.1  CCTTTTTGCTCCCCACGCTTTCGTGCAATGAGAGTCAGTAATGCGC...\n",
      "RS_GCF_900130015.1  GGAGAGTTTGATCCTGGCTCAGAACGAACGCTGGCGGCGTGCTTAA...\n",
      "GB_GCA_003288255.1  TTGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "\n",
    "# Set the path to your FASTA file\n",
    "fasta_file = \"data/ready4train_seqs.fasta\"\n",
    "\n",
    "# Parse the sequences and headers from the FASTA file\n",
    "records = SeqIO.parse(fasta_file, \"fasta\")\n",
    "\n",
    "# Create a DataFrame from the sequences and headers\n",
    "df = pd.DataFrame([(record.id, str(record.seq)) for record in records],\n",
    "                  columns=[\"id\", \"sequence\"])\n",
    "df = df.set_index('id')\n",
    "# Print the first 10 rows of the DataFrame\n",
    "print(df.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3a63bf08-cc2b-460c-ac59-5869d85c2dc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>domain</th>\n",
       "      <th>phylum</th>\n",
       "      <th>class</th>\n",
       "      <th>order</th>\n",
       "      <th>family</th>\n",
       "      <th>genus</th>\n",
       "      <th>species</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Seq_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RS_GCF_003697165.2</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Escherichia</td>\n",
       "      <td>Escherichia coli</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001027105.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Staphylococcales</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcus aureus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000006945.2</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>Salmonella enterica</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000742135.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Enterobacterales</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001457635.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Streptococcaceae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000195955.2</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Actinobacteriota</td>\n",
       "      <td>Actinomycetia</td>\n",
       "      <td>Mycobacteriales</td>\n",
       "      <td>Mycobacteriaceae</td>\n",
       "      <td>Mycobacterium</td>\n",
       "      <td>Mycobacterium tuberculosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001457615.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonas aeruginosa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_009759685.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Gammaproteobacteria</td>\n",
       "      <td>Pseudomonadales</td>\n",
       "      <td>Moraxellaceae</td>\n",
       "      <td>Acinetobacter</td>\n",
       "      <td>Acinetobacter baumannii</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001077535.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes_A</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Peptostreptococcales</td>\n",
       "      <td>Peptostreptococcaceae</td>\n",
       "      <td>Clostridioides</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001544255.1</th>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Bacilli</td>\n",
       "      <td>Lactobacillales</td>\n",
       "      <td>Enterococcaceae</td>\n",
       "      <td>Enterococcus_B</td>\n",
       "      <td>Enterococcus_B faecium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      domain            phylum                class  \\\n",
       "Seq_ID                                                                \n",
       "RS_GCF_003697165.2  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "RS_GCF_001027105.1  Bacteria        Firmicutes              Bacilli   \n",
       "RS_GCF_000006945.2  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "RS_GCF_000742135.1  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "RS_GCF_001457635.1  Bacteria        Firmicutes              Bacilli   \n",
       "RS_GCF_000195955.2  Bacteria  Actinobacteriota        Actinomycetia   \n",
       "RS_GCF_001457615.1  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "RS_GCF_009759685.1  Bacteria    Proteobacteria  Gammaproteobacteria   \n",
       "RS_GCF_001077535.1  Bacteria      Firmicutes_A           Clostridia   \n",
       "RS_GCF_001544255.1  Bacteria        Firmicutes              Bacilli   \n",
       "\n",
       "                                   order                 family  \\\n",
       "Seq_ID                                                            \n",
       "RS_GCF_003697165.2      Enterobacterales     Enterobacteriaceae   \n",
       "RS_GCF_001027105.1      Staphylococcales      Staphylococcaceae   \n",
       "RS_GCF_000006945.2      Enterobacterales     Enterobacteriaceae   \n",
       "RS_GCF_000742135.1      Enterobacterales     Enterobacteriaceae   \n",
       "RS_GCF_001457635.1       Lactobacillales       Streptococcaceae   \n",
       "RS_GCF_000195955.2       Mycobacteriales       Mycobacteriaceae   \n",
       "RS_GCF_001457615.1       Pseudomonadales       Pseudomonadaceae   \n",
       "RS_GCF_009759685.1       Pseudomonadales          Moraxellaceae   \n",
       "RS_GCF_001077535.1  Peptostreptococcales  Peptostreptococcaceae   \n",
       "RS_GCF_001544255.1       Lactobacillales        Enterococcaceae   \n",
       "\n",
       "                             genus                     species  \n",
       "Seq_ID                                                          \n",
       "RS_GCF_003697165.2     Escherichia            Escherichia coli  \n",
       "RS_GCF_001027105.1  Staphylococcus       Staphylococcus aureus  \n",
       "RS_GCF_000006945.2      Salmonella         Salmonella enterica  \n",
       "RS_GCF_000742135.1      Klebsiella       Klebsiella pneumoniae  \n",
       "RS_GCF_001457635.1   Streptococcus    Streptococcus pneumoniae  \n",
       "RS_GCF_000195955.2   Mycobacterium  Mycobacterium tuberculosis  \n",
       "RS_GCF_001457615.1     Pseudomonas      Pseudomonas aeruginosa  \n",
       "RS_GCF_009759685.1   Acinetobacter     Acinetobacter baumannii  \n",
       "RS_GCF_001077535.1  Clostridioides    Clostridioides difficile  \n",
       "RS_GCF_001544255.1  Enterococcus_B      Enterococcus_B faecium  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "taxonomy = pd.read_csv('data/train_taxonomy.tsv', sep='\\t', index_col = 0)\n",
    "taxonomy.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "931a4d2b-d7a0-4436-a080-6a042164d938",
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy['sequence'] = df['sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ef624bfb-c14f-4e2b-87d2-b83f2fe422d0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                      domain              phylum                class  \\\n",
      "Seq_ID                                                                  \n",
      "RS_GCF_019049245.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_000772045.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_004117935.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_010550605.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_002754715.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "...                      ...                 ...                  ...   \n",
      "RS_GCF_003852705.1  Bacteria        Bacteroidota          Bacteroidia   \n",
      "RS_GCF_018135625.1  Bacteria      Proteobacteria  Alphaproteobacteria   \n",
      "RS_GCF_000429325.1  Bacteria  Desulfobacterota_I     Desulfovibrionia   \n",
      "GB_GCA_017546175.1  Bacteria     Elusimicrobiota        Elusimicrobia   \n",
      "GB_GCA_002337145.1  Bacteria        Bacteroidota          Bacteroidia   \n",
      "\n",
      "                                 order               family  \\\n",
      "Seq_ID                                                        \n",
      "RS_GCF_019049245.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_000772045.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_004117935.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_010550605.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_002754715.1    Streptomycetales    Streptomycetaceae   \n",
      "...                                ...                  ...   \n",
      "RS_GCF_003852705.1    Flavobacteriales        Weeksellaceae   \n",
      "RS_GCF_018135625.1     Caulobacterales     Caulobacteraceae   \n",
      "RS_GCF_000429325.1  Desulfovibrionales  Desulfovibrionaceae   \n",
      "GB_GCA_017546175.1    Elusimicrobiales    Elusimicrobiaceae   \n",
      "GB_GCA_002337145.1     Chitinophagales     Chitinophagaceae   \n",
      "\n",
      "                                 genus                       species  \\\n",
      "Seq_ID                                                                 \n",
      "RS_GCF_019049245.1        Streptomyces      Streptomyces sp019049245   \n",
      "RS_GCF_000772045.1        Streptomyces      Streptomyces sp000772045   \n",
      "RS_GCF_004117935.1        Streptomyces   Streptomyces roseicoloratus   \n",
      "RS_GCF_010550605.1        Streptomyces      Streptomyces sp010550605   \n",
      "RS_GCF_002754715.1        Streptomyces      Streptomyces sp002754715   \n",
      "...                                ...                           ...   \n",
      "RS_GCF_003852705.1   Amniculibacterium   Amniculibacterium aquaticum   \n",
      "RS_GCF_018135625.1             BOG-935           BOG-935 sp018135625   \n",
      "RS_GCF_000429325.1  Fundidesulfovibrio   Fundidesulfovibrio putealis   \n",
      "GB_GCA_017546175.1    Avelusimicrobium  Avelusimicrobium sp017546175   \n",
      "GB_GCA_002337145.1             UBA1930           UBA1930 sp002337145   \n",
      "\n",
      "                                                             sequence  \n",
      "Seq_ID                                                                 \n",
      "RS_GCF_019049245.1  AGGAGGTGATCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...  \n",
      "RS_GCF_000772045.1  CGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...  \n",
      "RS_GCF_004117935.1  CGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...  \n",
      "RS_GCF_010550605.1  AGGAGGTGATCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...  \n",
      "RS_GCF_002754715.1  GGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTAA...  \n",
      "...                                                               ...  \n",
      "RS_GCF_003852705.1  GCGGCGTACGGGTGCGTAACACGTGTGCAACCTGCCTTTATCTAGG...  \n",
      "RS_GCF_018135625.1  AGGAGGTGATCCAGCCGCAGGTTCCCCTACGGCTACCTTGTTACGA...  \n",
      "RS_GCF_000429325.1  AGGAGGTGATCCAGCCGCAGGTTCCCCTACGGCTACCTTGTTACGA...  \n",
      "GB_GCA_017546175.1  GGAGAGTTTGATCCTGGCTCAGAGTTAACGCTGGCATCGTGCATAA...  \n",
      "GB_GCA_002337145.1  AGGAGGTATTCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...  \n",
      "\n",
      "[15265 rows x 8 columns]\n",
      "                      domain              phylum                class  \\\n",
      "Seq_ID                                                                  \n",
      "RS_GCF_000993785.3  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_018638295.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_008042045.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_003949795.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "RS_GCF_014648855.1  Bacteria    Actinobacteriota        Actinomycetia   \n",
      "...                      ...                 ...                  ...   \n",
      "GB_GCA_018266965.1  Bacteria        Bacteroidota          Bacteroidia   \n",
      "GB_GCA_003165275.1  Bacteria      Proteobacteria  Alphaproteobacteria   \n",
      "RS_GCF_013019105.1  Bacteria  Desulfobacterota_I     Desulfovibrionia   \n",
      "GB_GCA_017520185.1  Bacteria     Elusimicrobiota        Elusimicrobia   \n",
      "GB_GCA_018061385.1  Bacteria        Bacteroidota          Bacteroidia   \n",
      "\n",
      "                                 order               family  \\\n",
      "Seq_ID                                                        \n",
      "RS_GCF_000993785.3    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_018638295.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_008042045.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_003949795.1    Streptomycetales    Streptomycetaceae   \n",
      "RS_GCF_014648855.1    Streptomycetales    Streptomycetaceae   \n",
      "...                                ...                  ...   \n",
      "GB_GCA_018266965.1    Flavobacteriales        Weeksellaceae   \n",
      "GB_GCA_003165275.1     Caulobacterales     Caulobacteraceae   \n",
      "RS_GCF_013019105.1  Desulfovibrionales  Desulfovibrionaceae   \n",
      "GB_GCA_017520185.1    Elusimicrobiales    Elusimicrobiaceae   \n",
      "GB_GCA_018061385.1     Chitinophagales     Chitinophagaceae   \n",
      "\n",
      "                                 genus                         species  \\\n",
      "Seq_ID                                                                   \n",
      "RS_GCF_000993785.3        Streptomyces        Streptomyces xiamenensis   \n",
      "RS_GCF_018638295.1        Streptomyces        Streptomyces sp018638295   \n",
      "RS_GCF_008042045.1        Streptomyces        Streptomyces sp008042045   \n",
      "RS_GCF_003949795.1        Streptomyces        Streptomyces sp003949795   \n",
      "RS_GCF_014648855.1        Streptomyces           Streptomyces parvulus   \n",
      "...                                ...                             ...   \n",
      "GB_GCA_018266965.1   Amniculibacterium   Amniculibacterium sp018266965   \n",
      "GB_GCA_003165275.1             BOG-935             BOG-935 sp003165275   \n",
      "RS_GCF_013019105.1  Fundidesulfovibrio  Fundidesulfovibrio sp013019105   \n",
      "GB_GCA_017520185.1    Avelusimicrobium    Avelusimicrobium sp017520185   \n",
      "GB_GCA_018061385.1             UBA1930             UBA1930 sp018061385   \n",
      "\n",
      "                                                             sequence  \n",
      "Seq_ID                                                                 \n",
      "RS_GCF_000993785.3  TGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...  \n",
      "RS_GCF_018638295.1  AGGAGGTGATCCAGCCGCACCTTCCGGTACGGCTACCTTGTTACGA...  \n",
      "RS_GCF_008042045.1  TGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...  \n",
      "RS_GCF_003949795.1  CGGAGAGTTTGATCCTGGCTCAGGACGAACGCTGGCGGCGTGCTTA...  \n",
      "RS_GCF_014648855.1  TTGTTACGACTTCGTCCCAATCGCCAGTCCCACCTTCGACAGCTCC...  \n",
      "...                                                               ...  \n",
      "GB_GCA_018266965.1  GATCCTTCCGCTGGTTCACCAACGGAAACCTTGTTACGACTTCTCC...  \n",
      "GB_GCA_003165275.1  TGAGAGTTTGATCCTGGCTCNNGAGCGAACGCTGGCGGCAGGCCTA...  \n",
      "RS_GCF_013019105.1  AGGAGGTGATCCAGCCGCAGGTTCCCCTACGGCTACCTTGTTACGA...  \n",
      "GB_GCA_017520185.1  TGGAGAGTTTGATCCTGGCTCAGAGTTAACGCTGGCATCGTGCATA...  \n",
      "GB_GCA_018061385.1  CCATTGTCCAATATTCCTTACTGCTGCCTCCCGTAGGAGTCGGGCC...  \n",
      "\n",
      "[17036 rows x 8 columns]\n"
     ]
    }
   ],
   "source": [
    "# Get rows with at least two occurrences of the value in the \"genus\" column\n",
    "genus_counts = taxonomy['genus'].value_counts()\n",
    "genus_multiple_occurrences = genus_counts[genus_counts >= 2].index.tolist()\n",
    "df_subset = taxonomy[taxonomy['genus'].isin(genus_multiple_occurrences)]\n",
    "\n",
    "# Split the DataFrame into two subsets, test and ref, with about 50% of each genus value\n",
    "test = pd.DataFrame()\n",
    "ref = pd.DataFrame()\n",
    "for genus in genus_multiple_occurrences:\n",
    "    genus_subset = df_subset[df_subset['genus'] == genus]\n",
    "    n = len(genus_subset)\n",
    "    indices = np.arange(n)\n",
    "    np.random.shuffle(indices)\n",
    "    split_idx = int(np.floor(n / 2))\n",
    "    test_idx = indices[:split_idx]\n",
    "    ref_idx = indices[split_idx:]\n",
    "    test = pd.concat([test, genus_subset.iloc[test_idx]])\n",
    "    ref = pd.concat([ref, genus_subset.iloc[ref_idx]])\n",
    "\n",
    "# Print the resulting DataFrames\n",
    "print(test)\n",
    "print(ref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "e16841be-aec3-48f5-b280-39a02c1a55eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = SentenceTransformer('all-MiniLM-L6-v2') # can i use a biological dna model from hugging face? "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "0d7f9b29-ae08-4928-a8a2-42ac674e93f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test.sequence[:100].tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "65db76cc-86e2-4b1e-86ac-d253fcf3307f",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'word_dict' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_11001/400870103.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     33\u001b[0m     \u001b[0;31m# Get all sublists where element 1 equals the match string\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m     \u001b[0mclosest_match\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0msublist\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0msublist\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msublist\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m     \u001b[0mcorrect_match\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mword_dict\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m     \u001b[0maccuracy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0many\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mclosest_match\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mitem\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcorrect_match\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'word_dict' is not defined"
     ]
    }
   ],
   "source": [
    "# Initialize variables for evaluation metrics\n",
    "total_accuracy = 0\n",
    "count = 0\n",
    "\n",
    "test_list = test.sequence[:100].tolist()\n",
    "ref_list = ref.sequence[:100].tolist()\n",
    "\n",
    "# Single list of sentences - Possible tens of thousands of sentences\n",
    "sequences = test_list + ref_list\n",
    "paraphrases = util.paraphrase_mining(model, sequences)  \n",
    "\n",
    "# Group sublists by their second element (index 1)\n",
    "grouped_dict = {}\n",
    "for sublist in paraphrases:\n",
    "    key = sublist[1]\n",
    "    if key not in grouped_dict:\n",
    "        grouped_dict[key] = []\n",
    "    grouped_dict[key].append(sublist)\n",
    "\n",
    "# Get the two sublists with the greatest value in element 0 for each unique value in element 1\n",
    "result = []\n",
    "for key in sorted(grouped_dict.keys()):\n",
    "    sublists = grouped_dict[key]\n",
    "    top_two_sublists = sorted(sublists, key=lambda x: x[0], reverse=True)[:2]\n",
    "    result.extend(top_two_sublists)\n",
    "\n",
    "# Replace indices with corresponding sentences\n",
    "for sublist in result:\n",
    "    sublist[1] = sequences[sublist[1]]\n",
    "    sublist[2] = sequences[sublist[2]]\n",
    "\n",
    "for seq in test:\n",
    "    # Get all sublists where element 1 equals the match string\n",
    "    closest_match = [sublist[2] for sublist in result if ((sublist[1] == seq))]\n",
    "    correct_match = [key for key, values in seq_dict.items() if seq in values] # make sequence dictionary\n",
    "    accuracy = int(any(item in closest_match for item in correct_match))\n",
    "    \n",
    "    # Update evaluation metrics\n",
    "    total_accuracy += accuracy\n",
    "    count += 1\n",
    "    \n",
    "    print(f\"Input taxa: {input_verb}, closest match: {closest_match}, correct match: {correct_match}\")\n",
    "\n",
    "# Compute average evaluation metrics\n",
    "avg_accuracy = total_accuracy / count\n",
    "\n",
    "print(f\"Average Accuracy: {avg_accuracy}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b93b5f13-f7c3-4401-83d3-10331d292e0f",
   "metadata": {},
   "source": [
    "### can also try below with DNA sequence model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a083d86e-6c50-41e9-892a-713645a10a80",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading checkpoint shards: 100%|██████████| 2/2 [00:29<00:00, 14.53s/it]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Embeddings shape: (1, 10, 2560)\n",
      "Embeddings per token: [[[-0.04825436  0.21411765  0.07392845 ...  0.05151267  0.01025398\n",
      "    0.21558578]\n",
      "  [-0.36657548  0.19614793  0.17926945 ...  0.42881718  0.10352972\n",
      "   -0.07980208]\n",
      "  [ 0.33798268  0.1696109  -0.03771623 ...  0.7363859  -0.519177\n",
      "    0.21427338]\n",
      "  ...\n",
      "  [ 0.30988532  0.47271636 -0.08401014 ...  0.7210888  -0.69447356\n",
      "    0.08499449]\n",
      "  [ 0.19159164  0.52927226 -0.13266128 ...  0.73496157 -0.63323313\n",
      "    0.02344994]\n",
      "  [ 0.20589665  0.3151195  -0.19496821 ...  0.6304233  -0.7212018\n",
      "    0.01103845]]]\n",
      "Mean sequence embeddings: tensor([[ 0.1267,  0.4275, -0.0254,  ...,  0.5978, -0.4886,  0.0630]])\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModelForMaskedLM\n",
    "import torch\n",
    "\n",
    "# Import the tokenizer and the model\n",
    "tokenizer = AutoTokenizer.from_pretrained(\"InstaDeepAI/nucleotide-transformer-2.5b-multi-species\")\n",
    "model = AutoModelForMaskedLM.from_pretrained(\"InstaDeepAI/nucleotide-transformer-2.5b-multi-species\")\n",
    "\n",
    "# Create a dummy dna sequence and tokenize it\n",
    "sequences = ['ATTCTG' * 9]\n",
    "tokens_ids = tokenizer.batch_encode_plus(sequences, return_tensors=\"pt\")[\"input_ids\"]\n",
    "\n",
    "# Compute the embeddings\n",
    "attention_mask = tokens_ids != tokenizer.pad_token_id\n",
    "torch_outs = model(\n",
    "    tokens_ids,\n",
    "    attention_mask=attention_mask,\n",
    "    encoder_attention_mask=attention_mask,\n",
    "    output_hidden_states=True\n",
    ")\n",
    "\n",
    "# Compute sequences embeddings\n",
    "embeddings = torch_outs['hidden_states'][-1].detach().numpy()\n",
    "print(f\"Embeddings shape: {embeddings.shape}\")\n",
    "print(f\"Embeddings per token: {embeddings}\")\n",
    "\n",
    "# Compute mean embeddings per sequence\n",
    "mean_sequence_embeddings = torch.sum(attention_mask.unsqueeze(-1)*embeddings, axis=-2)/torch.sum(attention_mask, axis=-1)\n",
    "print(f\"Mean sequence embeddings: {mean_sequence_embeddings}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2b6a34-2532-455f-8c19-60debd8259bb",
   "metadata": {
    "tags": []
   },
   "source": [
    "### need to get data andfind method to compare label to true or false label, or I can try to get database from green genes and then compare the ability of this method to the classification of greengenes using a real data set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "37aabea4-e337-4664-a17e-1a2ed85ac656",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['ATTCTGATTCTGATTCTGATTCTGATTCTGATTCTGATTCTGATTCTGATTCTG']\n"
     ]
    }
   ],
   "source": [
    "print(sequences)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "973066cf-a0df-4e35-a3e8-9106a658c0f9",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio_nlp",
   "language": "python",
   "name": "bio_nlp"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
